File: src/main/java/icu/takeneko/omms/central/network/session/response/Response.java

package icu.takeneko.omms.central.network.session.response;

import com.google.gson.GsonBuilder;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;

public class Response {
    private @NotNull Result responseCode;
    private Map<String, String> content;

    public Response(@NotNull Result result, Map<String, String> content) {
        this.responseCode = result;
        this.content = content;
    }

    public Response() {
        super();
        this.content = new HashMap<>();
        this.responseCode = Result.OK;
    }

    public static String serialize(Response response) {
        return (new GsonBuilder()).serializeNulls().create().toJson(response);
    }

    public static Response deserialize(String x) {
        return (new GsonBuilder()).serializeNulls().create().fromJson(x, Response.class);
    }

    public @NotNull Result getResponseCode() {
        return this.responseCode;
    }

    public void setResponseCode(@NotNull Result responseCode) {
        this.responseCode = responseCode;
    }

    public Map<String, String> getContent() {
        return this.content;
    }

    public void setContent(Map<String, String> content) {
        this.content = content;
    }

    public @NotNull Response withResponseCode(@NotNull Result code) {
        this.setResponseCode(code);
        return this;
    }

    public @NotNull Response withContentPair(String a, String b) {
        this.content.put(a, b);
        return this;
    }

    public @NotNull String toString() {

        return "Response{code='" + this.responseCode + "', content=" + this.content + "}";
    }
}

File: src/main/java/icu/takeneko/omms/central/network/session/response/Response.kt
package icu.takeneko.omms.central.network.session.response

import icu.takeneko.omms.central.network.session.FailureReasons
import icu.takeneko.omms.central.util.Util

@kotlinx.serialization.Serializable
data class Response(
    val requestId: String,
    var event: Status,
    var content: MutableMap<String, String>
) {

    fun withContentPair(a: String, b: String): Response {
        content[a] = b
        return this
    }

    fun withMark(m: String): Response {
        content["marker_$m"] = ""
        return this;
    }

    fun withFailureReason(b: String): Response {
        content["reason"] = b
        return this
    }


    fun withContentPair(a: String, b: Any): Response = withContentPair(a, Util.toJson(b))

    operator fun set(key: String, value: String) {
        withContentPair(key, value)
    }

    operator fun contains(key: String): Boolean {
        return this.content.containsKey(key)
    }

    companion object {
        fun rateExceeded() = Response("", Status.FAIL, mutableMapOf("reason" to FailureReasons.RATE_EXCEED))
    }

}